networkx/utils/decorators.py

```diff
@@ -392,84 +355,868 @@ def random_array(dims, random_state=1):
+    Let's start with a caveat. Most of these examples will use `@argmap(...)`
+    to apply the decorator immediately to the function defined on the next line.
+    In the NetworkX codebase however, `argmap` is used within a function to
+    construct a decorator. That is, the decorator defines a mapping function
+    and then uses `argmap` to build the decorator from the mapping.
+    A simple example is a decorator that specifies which currency to report money.
+    The decorator (named `convert_to`) would be used like::
+
+        @convert_to("US_Dollars", "income")
+        def show_me_the_money(name, income):
+            print(f"{name} : {income}")
+
+    And the code to create the decorator might be::
+
+        def convert_to(currency, which_arg):
+            def _convert(amount):
+                if amount.currency != currency:
+                    amount = amount.to_currency(currency)
+                return amount
+            return argmap(_convert, which_arg)
+
+    Despite this common idiom for argmap, most of the following examples
+    use the `@argmap(...)` idiom to save space here.
+
+    As an example, the decorated function::
+
+        @argmap(sum, "x", "z")
+        def foo(x, y, z):
+            return x - y + z
+
+    is syntactic sugar for::
 
-        # Create a numpy.random.RandomState instance
-        random_state = create_py_random_state(random_state_arg)
+        def foo(x, y, z):
+            x = sum(x)
+            z = sum(z)
+            return x - y + z
 
-        # args is a tuple, so we must convert to list before modifying it.
-        new_args = list(args)
-        new_args[random_state_index] = random_state
-        return func(*new_args, **kwargs)
+    and is equivalent to (using argument indexes)::
 
-    return _random_state
+        @argmap(sum, "x", 2)
+        def foo(x, y, z):
+            return x - y + z
+
+    or::
+
+        @argmap(sum, "z", 0)
+        def foo(x, y, z):
+            return x - y + z
+
+    Transforming functions can be applied to multiple arguments, such as::
+
+        def swap(x, y):
+            return y, x
+
+        # the 2-tuple tells argmap that the map `swap` has 2 inputs/outputs.
+        @argmap(swap, ("a", "b")):
+        def foo(a, b, c):
+            return a / b * c
+
+    is equivalent to::
+
+        def foo(a, b, c):
+            a, b = swap(a, b)
+            return a / b * c
+
+    More generally, the applied arguments can be nested tuples of strings or ints.
+    The syntax `@argmap(some_func, ("a", ("b", "c")))` would expect `some_func` to
+    accept 2 inputs with the second expected to be a 2-tuple. It should then return
+    2 outputs with the second a 2-tuple. The returns values would replace input "a"
+    "b" and "c" respectively. Similarly for `@argmap(some_func, (0, ("b", 2)))`.
+
+    Also, note that an index larger than the number of named parameters is allowed
+    so long as a VAR_POSITIONAL input appears, e.g. `*args`. In that case the index
```

Comment:
What is VAR_POSITIONAL? Will most readers know?

Reply:
That's elaborated at the end of the sentence: "e.g., `*args`".  Is that too unclear?  Can you think of a better way to phrase this?  I read a ton of documentation, tutorials, etc. about the `*args` and `**kwargs` syntax in function calls & definitions, and there doesn't seem to be a consistent way of talking about those argument containers.

Reply:
I just had never heard the word VAR_POSITIONAL, and presume others wouldn't have had either.  Are there others than `*args`?

I think these are officially called "variadic arguments" or variable length arguments.